src/tools/db-upgrade/DbUpgrader.cpp

```diff
@@ -0,0 +1,504 @@
+                auto dvId = NebulaKeyUtilsV1::getDstId(key);
+                auto version = NebulaKeyUtilsV1::getVersion(key);
+
+                auto it = edgeSchemas_.find(std::abs(edgetype));
+                if (it == edgeSchemas_.end()) {
+                    // Invalid data
+                    iter->next();
+                    continue;
+                }
+                if (svId == lastSrcVertexId &&
+                    edgetype == lastEdgeType &&
+                    ranking == lastRank &&
+                    dvId == lastDstVertexId) {
+                    // Multi version
+                    iter->next();
+                    continue;
+                }
+
+                // auto strsvId = folly::to<std::string>(svId);
+                // auto strdvId = folly::to<std::string>(dvId);
+                auto strsvId = std::string(reinterpret_cast<const char*>(&svId), sizeof(svId));
+                auto strdvId = std::string(reinterpret_cast<const char*>(&dvId), sizeof(dvId));
+                if (!isValidVidLen(strsvId, strdvId)) {
+                    iter->next();
+                    continue;
+                }
+
+                auto newEdgeSchema = it->second.back();
+
+                // Generate 2.0 key
+                auto newKey = NebulaKeyUtils::edgeKey(spaceVidLen_,
+                                                      partId,
+                                                      strsvId,
+                                                      edgetype,
+                                                      ranking,
+                                                      strdvId,
+                                                      version);
+                auto val = iter->val();
+                auto reader = RowReaderWrapper::getEdgePropReader(schemaMan_.get(),
+                                                                  spaceId_,
+                                                                  std::abs(edgetype),
+                                                                  val);
+                if (!reader) {
+                    LOG(INFO) << "Can't get edge reader of " << edgetype;
+                    iter->next();
+                    continue;
+                }
+                // Generate 2.0 value and index records
+                encodeEdgeValue(partId, reader.get(), newEdgeSchema.get(), newKey, strsvId,
+                                edgetype, ranking, strdvId, data);
+                lastSrcVertexId = svId;
+                lastEdgeType  = edgetype;
+                lastRank = ranking;
+                lastDstVertexId = dvId;
+            }
+
+            if (data.size() >= FLAGS_write_batch_num) {
+                LOG(INFO) << "Send record total rows " << data.size();
+                auto code = engine_->multiPut(data);
+                if (code != kvstore::ResultCode::SUCCEEDED) {
+                    LOG(ERROR)  << "Write multi put failed！";
```

Comment:
Caused data loss， is better to abort the upgrade？

Reply:
Good catch，fixed，thx